### DISCUSSION TITLE
Hidden DPS values, only displaying on refresh

### DISCUSSION THREAD

**Discussion by ghost** [codetheweb/tuyapi#518] [General]
I have two `IoT A61 RGB` and `IoT C1 RGB` bulbs, both from Hiper. I am facing an issue where DPS are not displayed on get data, although they are displayed on refresh. This is DPS from 27 to 32. And another problem is that one status is hidden in debug (this is DPS 27). Another problem is that DPS starts at 20, not 0.

```js
let data = {
  dps: {
    '20': false,
    '21': 'white',
    '22': 103,
    '23': 88,
    '24': '000003e803e8',
    '25': '000e0d0000000000000000c80000',
    '26': 0
    //'27' is problem, it unknown
    // should to be have '28', '29', '30', '31', '32'
  }
}
```

Those hidden DPS... 

- `27` (hidden, unknown),
- `28` (music_data, empty),
- `29` (control_data, empty),
- `30` (rhythm_mode, base64),
- `31` (sleep_mode, base64),
- `32` (wakeup_mode, base64)

**Comment by codetheweb** [comment]
1. > DPS are not displayed on get data, although they are displayed on refresh

	I've seen this before, just seems to be an oddity of some devices.
2. > one status is hidden in debug (this is DPS 27).

	Again, device oddity.

3. > Another problem is that DPS starts at 20, not 0.

	See above. 😛 

I'm not seeing any issues that are specifically TuyAPI-related (if I misinterpreted something, please let me know). Converting this to a discussion for now.

**Comment by codetheweb** [comment]
See https://github.com/codetheweb/tuyapi/issues/510 for further context.

**Comment by RealIndica** [comment]
Hey, if you go on the tuya IOT website, login then : 
- Go to product 
- Development
- Create

then fill in the form and make sure you find the product that matches whatever you have then create it with a random name.
On the next page you should see a list of predefined DPS functions, they can guide you to what the DPS is on your device since the people who release their tuya products are lazy when it comes to development (most likely).

But basically, all this is that the DPS function has two ways to act.
- Send Only
- Report Only
- Send & Report

When a DPS function is set to Send Only, it will accept you using that DPS but will not show it in the schema.
When it is set to Send & Report, it will do as above and show it in the schema also.
When it is set to Report Only, the DPS value cannot be modified by the end user.

I recently tested this by running Tuya on my Raspberry Pi and hooking it up to my coffee machine (because i needed coffees to wake me up in the mornings and its brilliant lmao) and I can confirm that some DPS values are hidden because of them being set to send only.

There might be a way to reveal them, but for now I have no idea.
